File: packages/python/prompt_sanitizer/session.py
"""
Session — bidirectional anonymize/deanonymize for multi-turn LLM workflows.

A Session holds a shared Vault so that:
- The same original PII value always maps to the same token within a session.
- LLM responses can be deanonymized back to the originals.

Usage::

    s = Sanitizer(mode=Mode.SMART)

    session = s.session()
    clean_prompt = session.anonymize(user_prompt)
    llm_response = call_llm(clean_prompt)
    final = session.deanonymize(llm_response)

    # Context manager — vault cleared on exit
    with s.session() as sess:
        clean = sess.anonymize(text)
        ...

By default a session's vault lives only in process memory. Pass a
``store`` (see :mod:`prompt_sanitizer.vault_store`) to reattach to the same
mapping later — e.g. after a process restart — by ``session_id``::

    store = SQLiteVaultStore("./vault.db")
    session = s.session(session_id="user-42", store=store)
    clean = session.anonymize(user_prompt)
    session.persist()
    # ...later, possibly in a new process:
    resumed = s.session(session_id="user-42", store=store)
    final = resumed.deanonymize(llm_reply)
"""
from __future__ import annotations

from typing import TYPE_CHECKING, Optional

from .exceptions import VaultStoreError
from .vault import Vault
from .vault_store import BaseVaultStore, assert_supported_version, to_vault_snapshot

if TYPE_CHECKING:
    from .result import SanitizeResult
    from .sanitizer import Sanitizer


class Session:
    """
    Maintains a shared :class:`Vault` across multiple anonymize/deanonymize calls.

    Parameters
    ----------
    sanitizer:
        The parent :class:`Sanitizer` instance.
    session_id:
        Optional identifier included in audit log events, and required to
        use ``store``.
    store:
        Optional :class:`BaseVaultStore`. If given (with ``session_id``),
        any previously-persisted vault for this session is loaded
        synchronously before the session is returned to the caller.
    auto_persist:
        If True, persist to ``store`` at the end of every ``anonymize()``
        call. Default False — call :meth:`persist` explicitly.
    """

    def __init__(
        self,
        sanitizer: "Sanitizer",
        session_id: str | None = None,
        store: Optional[BaseVaultStore] = None,
        auto_persist: bool = False,
    ) -> None:
        self._sanitizer = sanitizer
        self._vault = Vault()
        self.session_id = session_id
        self._store = store
        self._auto_persist = auto_persist
        self._hydrate()

    def _hydrate(self) -> None:
        if self._store is None or self.session_id is None:
            return
        snapshot = self._store.load(self.session_id)
        if snapshot is None:
            return
        assert_supported_version(snapshot)
        self._vault.hydrate({"mappings": snapshot.mappings, "counters": snapshot.counters})

    # ── Core operations ──────────────────────────────────────────────────────

    def anonymize(self, text: str) -> str:
        """
        Sanitize *text*, returning the redacted version.

        Replacements are stored in the session vault for later deanonymization.
        """
        result = self._sanitizer._run(text, self._vault, session_id=self.session_id)
        if self._auto_persist:
            self.persist()
        return result.text

    def anonymize_with_result(self, text: str) -> "SanitizeResult":
        """Like :meth:`anonymize` but returns the full :class:`SanitizeResult`."""
        result = self._sanitizer._run(text, self._vault, session_id=self.session_id)
        if self._auto_persist:
            self.persist()
        return result

    def deanonymize(self, text: str) -> str:
        """
        Restore all known replacement tokens in *text* to their originals.

        Pass the LLM's response here to get a human-readable output with
        real names/values restored.
        """
        return self._vault.restore(text)

    def persist(self) -> None:
        """
        Persist the current vault state to this session's store.

        Raises :class:`VaultStoreError` if this session wasn't created with
        both a ``session_id`` and a ``store``.
        """
        if self._store is None or self.session_id is None:
            raise VaultStoreError(
                "Session.persist() requires both session_id and store to "
                "have been passed to Sanitizer.session()."
            )
        self._store.save(self.session_id, to_vault_snapshot(self.session_id, self._vault.to_data()))

    def forget(self) -> None:
        """
        Delete this session's persisted snapshot from its store, if any.
        Does not clear the in-memory vault — call :meth:`reset` for that too.
        """
        if self._store is None or self.session_id is None:
            return
        self._store.delete(self.session_id)

    def reset(self) -> None:
        """Clear the vault, starting a fresh mapping for this session."""
        self._vault.clear()

    # ── Introspection ────────────────────────────────────────────────────────

    @property
    def vault(self) -> Vault:
        """Direct access to the underlying vault (read-only recommended)."""
        return self._vault

    @property
    def mapping(self) -> dict[str, str]:
        """Snapshot of the current original → replacement mapping."""
        return self._vault.snapshot()

    def __len__(self) -> int:
        """Number of PII values currently stored in the session vault."""
        return len(self._vault)

    # ── Context manager ──────────────────────────────────────────────────────

    def __enter__(self) -> "Session":
        return self

    def __exit__(self, *_) -> None:
        self.reset()

    def __repr__(self) -> str:
        return f"Session(id={self.session_id!r}, mappings={len(self)})"

File: packages/ruby/lib/prompt_sanitizer/session.rb
# frozen_string_literal: true

module PromptSanitizer
  # Maintains a shared Vault across multiple anonymize/deanonymize calls.
  #
  # The same PII value always maps to the same replacement token within a
  # session, and LLM responses can be deanonymized back to the originals.
  #
  # Usage:
  #
  #   sanitizer = PromptSanitizer::Sanitizer.new(mode: :smart)
  #
  #   # Single-use block (vault cleared on exit):
  #   sanitizer.session(session_id: "user-42") do |sess|
  #     clean    = sess.anonymize(user_prompt)
  #     response = call_llm(clean)
  #     puts sess.deanonymize(response)
  #   end
  #
  #   # Long-lived (manage lifecycle yourself):
  #   sess = sanitizer.session
  #   loop { sess.anonymize(...) ... }
  #   sess.reset
  #
  # By default a session's vault lives only in process memory. Pass +store:+
  # (see PromptSanitizer::VaultStore) to reattach to the same mapping later
  # — e.g. after a process restart — by session_id:
  #
  #   store = PromptSanitizer::VaultStore::FileVaultStore.new("./vault-data")
  #   sess  = sanitizer.session(session_id: "user-42", store: store)
  #   clean = sess.anonymize(user_prompt)
  #   sess.persist
  #   # ...later, possibly in a new process:
  #   resumed = sanitizer.session(session_id: "user-42", store: store)
  #   final   = resumed.deanonymize(llm_reply)
  class Session
    attr_reader :session_id, :vault

    # @param sanitizer  [Sanitizer]
    # @param session_id [String, nil]
    # @param store [VaultStore::Base, nil] if given (with session_id), any
    #   previously-persisted vault for this session is loaded synchronously
    #   before the session is returned to the caller.
    # @param auto_persist [Boolean] if true, persist to +store+ at the end
    #   of every #anonymize call. Default false — call #persist explicitly.
    def initialize(sanitizer, session_id: nil, store: nil, auto_persist: false)
      @sanitizer    = sanitizer
      @session_id   = session_id
      @vault        = Vault.new
      @store        = store
      @auto_persist = auto_persist
      _hydrate
    end

    # Sanitize +text+ using the session's shared vault.
    # Returns the redacted string.
    #
    # @param text [String]
    # @return [String]
    def anonymize(text)
      result = @sanitizer._run(text, @vault, session_id: @session_id).text
      persist if @auto_persist
      result
    end

    # Like #anonymize but returns the full SanitizeResult.
    #
    # @param text [String]
    # @return [SanitizeResult]
    def anonymize_with_result(text)
      result = @sanitizer._run(text, @vault, session_id: @session_id)
      persist if @auto_persist
      result
    end

    # Persists the current vault state to this session's store.
    #
    # @raise [VaultStoreError] if this session wasn't created with both a
    #   session_id and a store.
    def persist
      unless @store && @session_id
        raise VaultStoreError,
              "Session#persist requires both session_id and store to have " \
              "been passed to Sanitizer#session."
      end

      @store.save(@session_id, VaultStore.to_snapshot(@session_id, @vault.to_data))
      nil
    end

    # Deletes this session's persisted snapshot from its store, if any.
    # Does not clear the in-memory vault — call #reset for that too.
    def forget
      return unless @store && @session_id

      @store.delete(@session_id)
      nil
    end

    # Restore vault tokens in +text+ back to the original PII values.
    # Pass the LLM's response here to get a human-readable output.
    #
    # @param text [String]
    # @return [String]
    def deanonymize(text)
      @vault.restore(text)
    end

    # Snapshot of original → replacement mapping for this session.
    #
    # @return [Hash{String => String}]
    def mapping
      @vault.snapshot
    end

    # Number of unique PII values currently stored in the session vault.
    #
    # @return [Integer]
    def size
      @vault.snapshot.size
    end

    # Clear the vault, resetting the session for a fresh conversation.
    def reset
      @vault.clear
    end

    # Block form — vault is cleared after the block returns.
    #
    #   sanitizer.session { |s| s.anonymize(...) }
    #
    # @yieldparam session [Session] self
    # @return the block's return value
    def use
      yield self
    ensure
      reset
    end

    def inspect
      "#<PromptSanitizer::Session id=#{@session_id.inspect} mappings=#{size}>"
    end
    alias to_s inspect

    private

    def _hydrate
      return unless @store && @session_id

      snapshot = @store.load(@session_id)
      return unless snapshot

      VaultStore.assert_supported_version!(snapshot)
      @vault.hydrate(mappings: snapshot.mappings, counters: snapshot.counters)
    end
  end
end
